File: xeg-upgrader/python/src/xeg_upgrader/xeg_reader.py
"""Load and validate an Evagene v1 ``.xeg`` XML file before upload.

Pure: no network, no logging.  The upload path depends on a guaranteed
well-formed document rooted at ``<Pedigree>``; that invariant belongs
here, not in the HTTP client.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from lxml import etree

_EXPECTED_ROOT = "Pedigree"
_UTF8_BOM = "\ufeff"


class InvalidXegError(ValueError):
    """Raised when a file is not a well-formed Evagene v1 ``.xeg``."""


@dataclass(frozen=True)
class XegDocument:
    raw_text: str


def read_from_file(path: str | Path) -> XegDocument:
    file_path = Path(path)
    if not file_path.is_file():
        raise InvalidXegError(f"file not found: {file_path}")
    try:
        text = file_path.read_text(encoding="utf-8")
    except OSError as exc:
        raise InvalidXegError(f"could not read {file_path}: {exc}") from exc
    return parse(text)


def parse(text: str) -> XegDocument:
    stripped = text.removeprefix(_UTF8_BOM)
    try:
        root = etree.fromstring(stripped.encode("utf-8"))
    except etree.XMLSyntaxError as exc:
        raise InvalidXegError(
            f"not well-formed XML: {exc.msg} — check the file is a genuine Evagene v1 .xeg"
        ) from exc

    tag = etree.QName(root.tag).localname
    if tag != _EXPECTED_ROOT:
        raise InvalidXegError(
            f"root element is <{tag}>, expected <{_EXPECTED_ROOT}> — "
            "is this an Evagene v1 .xeg file?"
        )
    return XegDocument(raw_text=stripped)

File: xeg-upgrader/dotnet/src/XegUpgrader/XegReader.cs
using System.Xml;
using System.Xml.Linq;

namespace XegUpgrader;

public sealed class InvalidXegException : Exception
{
    public InvalidXegException(string message) : base(message) { }

    public InvalidXegException(string message, Exception inner) : base(message, inner) { }
}

public sealed record XegDocument(string RawText);

/// <summary>
/// Loads an Evagene v1 .xeg file from disk and validates that it is
/// well-formed XML rooted at a <c>&lt;Pedigree&gt;</c> element before the
/// upload attempts it.  Pure: no network, no side-effects beyond reading
/// the file handed in.
/// </summary>
public static class XegReader
{
    private const string ExpectedRoot = "Pedigree";

    public static XegDocument ReadFromFile(string path)
    {
        if (!File.Exists(path))
        {
            throw new InvalidXegException($"file not found: {path}");
        }

        string text;
        try
        {
            text = File.ReadAllText(path);
        }
        catch (IOException e)
        {
            throw new InvalidXegException($"could not read {path}: {e.Message}", e);
        }

        return Parse(text);
    }

    public static XegDocument Parse(string text)
    {
        var stripped = StripBom(text);
        XDocument document;
        try
        {
            document = XDocument.Parse(stripped);
        }
        catch (XmlException e)
        {
            throw new InvalidXegException(
                $"not well-formed XML: {e.Message} — check the file is a genuine Evagene v1 .xeg",
                e);
        }

        var root = document.Root
            ?? throw new InvalidXegException("XML document has no root element");
        if (root.Name.LocalName != ExpectedRoot)
        {
            throw new InvalidXegException(
                $"root element is <{root.Name.LocalName}>, expected <{ExpectedRoot}> — is this an Evagene v1 .xeg file?");
        }

        return new XegDocument(stripped);
    }

    private static string StripBom(string text)
    {
        return text.Length > 0 && text[0] == '\uFEFF' ? text[1..] : text;
    }
}
